lib/puma/events.rb

```diff
@@ -80,51 +79,44 @@ def debug(str)
     # Write +str+ to +@stderr+
     #
     def error(str)
-      @stderr.puts format("ERROR: #{str}")
+      @debug_logger.error_dump(text: format("ERROR: #{str}"))
       exit 1
     end
 
     def format(str)
       formatter.call(str)
     end
 
+    # An HTTP connection error has occurred.
+    # +error+ a connection exception, +env+ the request
+    #
+    def connection_error(error, env, text="HTTP connection error")
+      @debug_logger.error_dump(error: error, env: env, text: text)
+    end
+
     # An HTTP parse error has occurred.
-    # +server+ is the Server object, +env+ the request, and +error+ a
+    # +env+ the request, and +error+ a
     # parsing exception.
     #
-    def parse_error(server, env, error)
-      @stderr.puts "#{Time.now}: HTTP parse error, malformed request " \
-        "(#{env[HTTP_X_FORWARDED_FOR] || env[REMOTE_ADDR]}#{env[REQUEST_PATH]}): " \
-        "#{error.inspect}" \
-        "\n---\n"
+    def parse_error(error, env)
+      @debug_logger.error_dump(error: error, env: env, text: 'HTTP parse error, malformed request', force: true)
```

Comment:
(This is more a comment on the concept and not this specific line)

"debug logger" and the "force: true" reads a bit strange to me... what about using different [levels](https://ruby-doc.org/stdlib-2.4.0/libdoc/logger/rdoc/Logger.html) instead? 

Just throwing this out as an idea, not demanding anything, this is already great work, thanks for doing it 👍

Reply:
Yeah, you're definitely right. It's strange and needs to be simplified, I'll have a look what I can do.

Reply:
I updated the MR with the new name and the new interface of the logger class.
In my opinion, now it looks better, but there is a couple of things that confuse me:

- `info` and `debug` methods prints headers and body of the request. As for me, this is ok for `debug`, but I'm not sure about `info`
- Since we replaced all `$stderr` invokes with the new interface, `log`, `write`, and `debug` use `$stdout` and have the same format, but `error` uses the new interface and have a different format

@nateberkopec what do you think about it?